In [1]:
%load_ext sql

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


'Connected: mildsalmon_su@dev'

# 1. Assignment 2 : 사용자별로 처음과 마지막 채널 알아내기

쿼리를 많이 날려봤는데, 어떤 것이 정답인지 모르겠다. 그래서 test case 10개 정도를 출력해보고 내가 작성한 쿼리와 비교해보려고 한다.


## A. test case

|user |first |last|
|---|---|---|
|27	| Youtube |  Instagram|
|29|	Naver	 |   Naver|
|33|	Google  |  Youtube|
|34	|Youtube|	  Naver|
|36|	Naver	 |   Youtube|
|40|	Youtube |  Google|
|41	|Facebook | Youtube|
|44|	Naver	|   Instagram|
|45|Youtube	|  Instagram|
|59	|Instagram	|Instagram|

### a. 사용자별 첫 번째 채널 (10명)

In [ ]:
%%sql

SELECT userid
     , channel
     , ts
     , num
FROM (
  SELECT usc.userid
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS num
       , usc.channel
       , st.ts
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE num = 1
ORDER BY 1, 3
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 1),
 (29, 'Naver', datetime.datetime(2019, 5, 1, 19, 13, 33, 147000), 1),
 (33, 'Google', datetime.datetime(2019, 5, 1, 12, 22, 45, 597000), 1),
 (34, 'Youtube', datetime.datetime(2019, 5, 1, 18, 10, 14, 367000), 1),
 (36, 'Naver', datetime.datetime(2019, 5, 1, 15, 7, 43, 370000), 1),
 (40, 'Youtube', datetime.datetime(2019, 5, 1, 14, 23, 7, 660000), 1),
 (41, 'Facebook', datetime.datetime(2019, 5, 1, 19, 49, 24, 53000), 1),
 (44, 'Naver', datetime.datetime(2019, 5, 2, 3, 34, 18, 413000), 1),
 (45, 'Youtube', datetime.datetime(2019, 5, 1, 15, 53, 7, 17000), 1),
 (59, 'Instagram', datetime.datetime(2019, 5, 1, 14, 1, 3, 87000), 1)]

### b. 사용자별 마지막 채널 (10명)

In [ ]:
%%sql

SELECT userid
     , channel
     , ts
     , num
FROM (
  SELECT usc.userid
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts DESC) AS num
       , usc.channel
       , st.ts
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE num = 1
ORDER BY 1, 3
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Instagram', datetime.datetime(2019, 11, 30, 22, 26, 35), 1),
 (29, 'Naver', datetime.datetime(2019, 11, 1, 19, 22, 39, 330000), 1),
 (33, 'Youtube', datetime.datetime(2019, 11, 30, 13, 15, 52), 1),
 (34, 'Naver', datetime.datetime(2019, 11, 28, 19, 5, 44), 1),
 (36, 'Youtube', datetime.datetime(2019, 7, 5, 14, 43, 49, 3000), 1),
 (40, 'Google', datetime.datetime(2019, 11, 30, 17, 58, 16), 1),
 (41, 'Youtube', datetime.datetime(2019, 11, 18, 19, 26, 35), 1),
 (44, 'Instagram', datetime.datetime(2019, 6, 16, 4, 0, 57, 210000), 1),
 (45, 'Instagram', datetime.datetime(2019, 11, 30, 22, 54, 15), 1),
 (59, 'Instagram', datetime.datetime(2019, 11, 30, 15, 13, 18), 1)]

### c. 위에서 구한 값이 맞는지 27번만 뽑아서 확인해보기

1. 아래처럼 노가다 방식으로도 확인하기

In [ ]:
%%sql

SELECT usc.userid
     , st.ts
     , usc.channel
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
WHERE usc.userid = 27
ORDER BY 1, 2
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
159 rows affected.


[(27, datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 'Youtube'),
 (27, datetime.datetime(2019, 5, 2, 19, 21, 30, 280000), 'Google'),
 (27, datetime.datetime(2019, 5, 3, 20, 38, 40, 730000), 'Naver'),
 (27, datetime.datetime(2019, 5, 4, 21, 48, 7, 300000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 5, 18, 15, 30, 540000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 6, 17, 49, 15, 437000), 'Google'),
 (27, datetime.datetime(2019, 5, 7, 20, 41, 25, 293000), 'Google'),
 (27, datetime.datetime(2019, 5, 8, 17, 35, 16, 440000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 9, 19, 57, 40, 37000), 'Organic'),
 (27, datetime.datetime(2019, 5, 10, 17, 57, 10, 207000), 'Organic'),
 (27, datetime.datetime(2019, 5, 11, 20, 23, 32, 920000), 'Organic'),
 (27, datetime.datetime(2019, 5, 11, 20, 28, 26, 313000), 'Organic'),
 (27, datetime.datetime(2019, 5, 12, 21, 24, 11, 560000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 12, 21, 32, 47, 7000), 'Naver'),
 (27, datetime.datetime(2019, 5, 13, 22, 15, 47, 763000), 'Naver'),
 (27, datetime.datetime(2019, 5, 14, 19, 48, 34, 197000), 'Organic'),
 (27, datetime.datetime(2019, 5, 15, 17, 19, 48, 473000), 'Google'),
 (27, datetime.datetime(2019, 5, 16, 18, 46, 44, 713000), 'Instagram'),
 (27, datetime.datetime(2019, 5, 17, 23, 30, 8, 503000), 'Google'),
 (27, datetime.datetime(2019, 5, 20, 21, 14, 17, 223000), 'Naver'),
 (27, datetime.datetime(2019, 5, 21, 19, 14, 31, 970000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 22, 19, 49, 35, 183000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 22, 19, 53, 19, 87000), 'Facebook'),
 (27, datetime.datetime(2019, 5, 24, 21, 55, 13, 513000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 4, 19, 2, 15, 837000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 5, 20, 11, 40, 597000), 'Google'),
 (27, datetime.datetime(2019, 6, 6, 18, 15, 33, 663000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 7, 22, 58, 7, 510000), 'Organic'),
 (27, datetime.datetime(2019, 6, 8, 17, 56, 34, 733000), 'Naver'),
 (27, datetime.datetime(2019, 6, 9, 18, 23, 17, 560000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 10, 19, 28, 31, 117000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 11, 17, 14, 19, 757000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 12, 17, 30, 43, 637000), 'Instagram'),
 (27, datetime.datetime(2019, 6, 14, 21, 32, 35, 627000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 15, 18, 59, 28, 643000), 'Organic'),
 (27, datetime.datetime(2019, 6, 21, 19, 8, 25, 137000), 'Organic'),
 (27, datetime.datetime(2019, 6, 22, 18, 13, 57, 810000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 23, 18, 44, 41, 623000), 'Organic'),
 (27, datetime.datetime(2019, 6, 24, 18, 13, 56, 720000), 'Google'),
 (27, datetime.datetime(2019, 6, 24, 18, 14, 10, 913000), 'Naver'),
 (27, datetime.datetime(2019, 6, 25, 20, 32, 25, 423000), 'Facebook'),
 (27, datetime.datetime(2019, 6, 26, 21, 6, 59, 173000), 'Naver'),
 (27, datetime.datetime(2019, 6, 27, 19, 20, 55, 307000), 'Youtube'),
 (27, datetime.datetime(2019, 6, 27, 19, 30, 30, 70000), 'Google'),
 (27, datetime.datetime(2019, 6, 28, 20, 10, 5, 257000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 8, 17, 31, 34, 597000), 'Organic'),
 (27, datetime.datetime(2019, 8, 9, 18, 35, 34, 240000), 'Organic'),
 (27, datetime.datetime(2019, 8, 9, 18, 36, 51, 883000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 10, 17, 54, 55, 827000), 'Facebook'),
 (27, datetime.datetime(2019, 8, 11, 20, 34, 47, 867000), 'Google'),
 (27, datetime.datetime(2019, 8, 12, 18, 18, 37, 337000), 'Youtube'),
 (27, datetime.datetime(2019, 8, 13, 19, 6, 44, 477000), 'Instagram'),
 (27, datetime.datetime(2019, 8, 14, 16, 16, 44, 857000), 'Naver'),
 (27, datetime.datetime(2019, 8, 15, 21, 26, 42, 373000), 'Organic'),
 (27, datetime.datetime(2019, 8, 16, 16, 7, 42, 797000), 'Google'),
 (27, datetime.datetime(2019, 8, 16, 16, 13, 53, 703000), 'Instagram'),
 (27, datetime.datetime(2019, 8, 17, 21, 47, 17, 573000), 'Naver'),
 (27, datetime.datetime(2019, 8, 18, 20, 54, 31

2. 아래처럼 row_number로 확인하기

In [ ]:
%%sql

SELECT userid
     , rank
     , ts
     , channel
FROM (
  SELECT usc.userid
      , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
      , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
      , st.ts
      , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE rank = 1 OR rank = cnt
ORDER BY 1
LIMIT 20
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
20 rows affected.


[(27, 1, datetime.datetime(2019, 5, 1, 17, 3, 59, 957000), 'Youtube'),
 (27, 159, datetime.datetime(2019, 11, 30, 22, 26, 35), 'Instagram'),
 (29, 1, datetime.datetime(2019, 5, 1, 19, 13, 33, 147000), 'Naver'),
 (29, 82, datetime.datetime(2019, 11, 1, 19, 22, 39, 330000), 'Naver'),
 (33, 1, datetime.datetime(2019, 5, 1, 12, 22, 45, 597000), 'Google'),
 (33, 222, datetime.datetime(2019, 11, 30, 13, 15, 52), 'Youtube'),
 (34, 1, datetime.datetime(2019, 5, 1, 18, 10, 14, 367000), 'Youtube'),
 (34, 99, datetime.datetime(2019, 11, 28, 19, 5, 44), 'Naver'),
 (36, 1, datetime.datetime(2019, 5, 1, 15, 7, 43, 370000), 'Naver'),
 (36, 108, datetime.datetime(2019, 7, 5, 14, 43, 49, 3000), 'Youtube'),
 (40, 1, datetime.datetime(2019, 5, 1, 14, 23, 7, 660000), 'Youtube'),
 (40, 245, datetime.datetime(2019, 11, 30, 17, 58, 16), 'Google'),
 (41, 1, datetime.datetime(2019, 5, 1, 19, 49, 24, 53000), 'Facebook'),
 (41, 90, datetime.datetime(2019, 11, 18, 19, 26, 35), 'Youtube'),
 (44, 1, datetime.datetime(2019, 5, 2, 3, 34, 18, 413000), 'Naver'),
 (44, 30, datetime.datetime(2019, 6, 16, 4, 0, 57, 210000), 'Instagram'),
 (45, 1, datetime.datetime(2019, 5, 1, 15, 53, 7, 17000), 'Youtube'),
 (45, 215, datetime.datetime(2019, 11, 30, 22, 54, 15), 'Instagram'),
 (59, 1, datetime.datetime(2019, 5, 1, 14, 1, 3, 87000), 'Instagram'),
 (59, 160, datetime.datetime(2019, 11, 30, 15, 13, 18), 'Instagram')]

## B. 풀이

### a. ROW_NUMBER()로 풀기

- 뭔가 억지로 맞춰서 푼 느낌이다.

|userID|rank|cnt|channel|
|---|---|---|---|
|27|1|159|Youtube|
|27|159|159|Instagram|

이런 방식으로 표현하는 것은 쉬웠는데, 2개의 row를 합치는 방법을 모르겠다.

#### 1) 정답

In [ ]:
%%sql

SELECT A.userid AS userID
     , A.first AS first_channel
     , B.last AS last_channel
FROM (
  SELECT userid
      , channel AS first
  FROM (
    SELECT usc.userid
        , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
        , usc.channel
    FROM raw_data.user_session_channel usc
    JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
  )
  WHERE rank = 1
) A JOIN (
  SELECT userid
      , channel AS last
  FROM (
    SELECT usc.userid
        , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
        , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
        , usc.channel
    FROM raw_data.user_session_channel usc
    JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
  )
  WHERE rank = cnt
) B ON (A.userid = B.userid)
ORDER BY 1
LIMIT 20
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
20 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram'),
 (64, 'Youtube', 'Youtube'),
 (65, 'Youtube', 'Organic'),
 (68, 'Youtube', 'Organic'),
 (69, 'Facebook', 'Instagram'),
 (80, 'Organic', 'Naver'),
 (84, 'Google', 'Youtube'),
 (87, 'Youtube', 'Google'),
 (97, 'Organic', 'Organic'),
 (112, 'Facebook', 'Youtube'),
 (113, 'Organic', 'Organic')]

#### 2) 시행착오

ROW_NUMBER()로만 푸는 방법은 없을까?

모르겠다..

In [ ]:
%%sql


SELECT userid
    , channel AS first
FROM (
  SELECT usc.userid
      , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
      , COUNT(1) OVER(PARTITION BY usc.userid) AS cnt
      , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
WHERE rank = 1 OR rank = cnt
GROUP BY userid
HAVING 
ORDER BY 1
LIMIT 20
;

### b. FIRST_VALUE / LAST_VALUE로 풀기

#### 1) 정답

In [ ]:
%%sql

SELECT DISTINCT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram')]

#### 2) 시행착오

아래 sql문은 PARTITION BY 뒤에 ORDER BY를 입력하지 않았기 때문에 정상적으로 출력되지 않는다.

In [ ]:
%%sql

SELECT DISTINCT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
5 rows affected.


[(27, 'Organic', 'Facebook'),
 (29, 'Facebook', 'Organic'),
 (33, 'Instagram', 'Youtube'),
 (34, 'Organic', 'Facebook'),
 (36, 'Facebook', 'Google')]

`ORDER BY`를 입력해주면 `Aggregate window functions with an ORDER BY clause require a frame clause` 에러가 발생한다.

`ORDER BY` 뒤에 세부적인 범위를 반드시 입력해줘야하는 걸까?

다른 윈도우 함수들은 `frame_clause`를 지정하지 않아도 잘 동작했기 때문에, `first_value`와 `last_value`만 세부적인 범위를 지정해줘야하는 것이 잘 이해가 가지 않았다.

그리고 그냥 넘어가기에는 찝찝해서 검색을 해봤다.

---

AWS Docs - FIRST_VALUE and LAST_VALUE window functions

- ORDER BY order_list

  - Sorts the rows within each partition. If no PARTITION BY clause is specified, ORDER BY sorts the entire table. If you specify an ORDER BY clause, you must also specify a frame_clause.
  - The results of the FIRST_VALUE and LAST_VALUE functions depend on the ordering of the data. The results are nondeterministic in the following cases:
  - When no ORDER BY clause is specified and a partition contains two different values for an expression
  - When the expression evaluates to different values that correspond to the same value in the ORDER BY list.

- frame_clause

  - If an ORDER BY clause is used for an aggregate function, an explicit frame clause is required. The frame clause refines the set of rows in a function's window, including or excluding sets of rows in the ordered result. The frame clause consists of the ROWS keyword and associated specifiers. See Window function syntax summary.


[1] AWS > Documentation > Amazon Redshift > Database Developer Guide. https://docs.aws.amazon.com/redshift/latest/dg/r_WF_first_value.html. (accessed Feb 20, 2022)

---

docs에 "`FIRST_VALUE`, `LAST_VALUE`에서 `ORDER BY`를 사용하기 위해서는 `frame_clause`를 명시해야한다고 한다."라고 적혀 있다.

In [ ]:
%%sql

SELECT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
(psycopg2.ProgrammingError) Aggregate window functions with an ORDER BY clause require a frame clause

[SQL: SELECT usc.userid
     , FIRST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
     , LAST_VALUE(usc.channel) OVER(PARTITION BY usc.userid ORDER BY st.ts)
FROM raw_data.user_session_channel usc
JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
ORDER BY 1
LIMIT 5
;]
(Background on this error at: https://sqlalche.me/e/14/f405)


### c. ROW_NUMBER()와 FIRST_VALUE / LAST_VALUE를 섞어서 풀기

풀고나서 생각했는데, B - b번 방식이 제일 깔끔한 것 같다.

#### 1) 정답

In [ ]:
%%sql

SELECT DISTINCT id
     , FIRST_VALUE(channel) OVER(PARTITION BY id ORDER BY rank ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
     , LAST_VALUE(channel) OVER(PARTITION BY id ORDER BY rank ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
FROM (
  SELECT usc.userid AS id
       , ROW_NUMBER() OVER(PARTITION BY usc.userid ORDER BY st.ts ASC) AS rank
       , usc.channel
  FROM raw_data.user_session_channel usc
  JOIN raw_data.session_timestamp st ON usc.sessionid = st.sessionid
)
ORDER BY 1
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(27, 'Youtube', 'Instagram'),
 (29, 'Naver', 'Naver'),
 (33, 'Google', 'Youtube'),
 (34, 'Youtube', 'Naver'),
 (36, 'Naver', 'Youtube'),
 (40, 'Youtube', 'Google'),
 (41, 'Facebook', 'Youtube'),
 (44, 'Naver', 'Instagram'),
 (45, 'Youtube', 'Instagram'),
 (59, 'Instagram', 'Instagram')]

# Assignment 3 : Gross Revenue가 가장 큰 UserID 10개 찾기

- Gross revenue
  - Refund 포함한 매출

## A. 풀이

In [ ]:
%%sql

SELECT B.userid AS userid
     , SUM(C.amount) AS gross_revenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.amount IS NOT NULL
GROUP BY B.userid
ORDER BY 2 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[(989, 743),
 (772, 556),
 (1615, 506),
 (654, 488),
 (1651, 463),
 (973, 438),
 (262, 422),
 (1099, 421),
 (2682, 414),
 (891, 412)]

## B. 검증

아래 SQL 결과값을 더해보면 위 풀이가 정답인지 알 수 있다.

In [ ]:
%%sql

SELECT B.userid AS userid
     , C.amount AS gross_revenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.amount IS NOT NULL
  AND userid IN (989, 772, 1615)
ORDER BY 1 DESC
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
19 rows affected.


[(1615, 62),
 (1615, 90),
 (1615, 109),
 (1615, 111),
 (1615, 70),
 (1615, 64),
 (989, 89),
 (989, 28),
 (989, 51),
 (989, 148),
 (989, 149),
 (989, 115),
 (989, 15),
 (989, 148),
 (772, 110),
 (772, 134),
 (772, 71),
 (772, 127),
 (772, 114)]

- sessionid당 구매가 1번만 발생했을까?
- 1번 접속한 사용자가 접속을 끊지 않고 물건을 2번 구매했을 경우는 없을까?
  - 없는것 같다.

In [ ]:
%%sql

SELECT sessionid
     , COUNT(sessionid)
FROM raw_data.session_transaction
GROUP BY sessionid
ORDER BY 2 DESC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('00029153d12ae1c9abe59c17ff2e0895', 1),
 ('008909bd27b680698322c750c400e4b1', 1),
 ('0107acb41ef20db2289d261d4e34fd38', 1),
 ('018544a2c48077d2cc3ce3278ab1e805', 1),
 ('020c38173caff02033564200e0288aa9', 1),
 ('029591145fad252b28d844d823097400', 1),
 ('03190309cbcd7161a426abd4782bdcd2', 1),
 ('03a13aa6921aed989b31d92dc73b391a', 1),
 ('04215c1ad3aaff3ed175446d2ea3c0f2', 1),
 ('049b8c570e6cc662b4a3d7e3c86e236f', 1)]

# Assignment 4 : 채널별 월 매출액 테이블 만들기

1. session_timestamp, user_session_channel, session_transaction 테이블들을 사용
2. 아래와 같은 필드로 구성
  - year-month
  - channel
  - uniqueUsers (총방문 사용자)
  - paidUsers (구매 사용자: refund한 경우도 판매로 고려) -> CASE WHEN 사용
  - conversionRate (구매사용자 / 총방문 사용자)
    - Casting이 필요. 분모는 NULLIF로 0으로 나누기 방지
  - grossRevenue (refund 포함)
  - netRevenue (refund 제외) -> CASE WHEN 사용


## A. 풀이




### a. 내 풀이

`D.refund`는 `SUM(C.amount)`로 refund 금액만 더한 column이다. 그래서 `MAX(D.refund)`를 사용해도 값이 달라지지 않는다. 그렇다고 `D.refund`를 사용하면 `GROUP BY`조건에 걸려서 에러가 발생한다.

---

근데 이게 최선일까?

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
     , B.channel AS channel
     , NULLIF(COUNT(DISTINCT B.userid), 0) AS uniqueUsers
     , CASE
        WHEN COUNT(amount) > 0 THEN COUNT(amount)
        WHEN COUNT(amount) = 0 THEN 0
       END AS paidUsers
     , paidUsers::float/NULLIF(uniqueUsers, 0) AS conversionRate
     , SUM(C.amount) AS grossRevenue
     , CASE
        WHEN MAX(D.refund) IS NULL THEN grossRevenue
        WHEN MAX(D.refund) IS NOT NULL THEN grossRevenue - MAX(D.refund)
       END AS netRevenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
LEFT JOIN (
  SELECT LEFT(A.ts, 7) AS yearMonth
       , B.channel AS channel
       , SUM(C.amount) AS refund
  FROM raw_data.session_timestamp A
  JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
  LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
  WHERE C.refunded IS TRUE
  GROUP BY 1, 2
) D ON (D.yearMonth = LEFT(A.ts, 7) AND D.channel = B.channel)
GROUP BY 1, 2
ORDER BY 1 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', 247, 14, 0.0566801619433198, 1199, 997),
 ('2019-05', 'Naver', 237, 11, 0.0464135021097046, 867, 844),
 ('2019-05', 'Organic', 238, 18, 0.0756302521008403, 1846, 1571),
 ('2019-05', 'Google', 253, 10, 0.0395256916996047, 580, 580),
 ('2019-05', 'Youtube', 244, 10, 0.040983606557377, 529, 529),
 ('2019-05', 'Instagram', 234, 11, 0.047008547008547, 959, 770),
 ('2019-06', 'Google', 412, 13, 0.0315533980582524, 947, 947),
 ('2019-06', 'Organic', 416, 14, 0.0336538461538462, 1129, 940),
 ('2019-06', 'Naver', 398, 16, 0.0402010050251256, 1090, 1090),
 ('2019-06', 'Instagram', 410, 21, 0.0512195121951219, 1462, 1418)]

In [ ]:
%%sql

DROP TABLE IF EXISTS mildsalmon_su.channel_month_gross_revenue;

CREATE TABLE mildsalmon_su.channel_month_gross_revenue AS
  SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , NULLIF(COUNT(DISTINCT B.userid), 0) AS uniqueUsers
      , CASE
          WHEN COUNT(amount) > 0 THEN COUNT(amount)
          WHEN COUNT(amount) = 0 THEN 0
        END AS paidUsers
      , paidUsers::float/NULLIF(uniqueUsers, 0) AS conversionRate
      , SUM(C.amount) AS grossRevenue
      , CASE
          WHEN MAX(D.refund) IS NULL THEN grossRevenue
          WHEN MAX(D.refund) IS NOT NULL THEN grossRevenue - MAX(D.refund)
        END AS netRevenue
  FROM raw_data.session_timestamp A
  JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
  LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
  LEFT JOIN (
    SELECT LEFT(A.ts, 7) AS yearMonth
        , B.channel AS channel
        , SUM(C.amount) AS refund
    FROM raw_data.session_timestamp A
    JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
    LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
    WHERE C.refunded IS TRUE
    GROUP BY 1, 2
  ) D ON (D.yearMonth = LEFT(A.ts, 7) AND D.channel = B.channel)
  GROUP BY 1, 2
  ORDER BY 1 ASC
  LIMIT 10;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
Done.
Done.


[]

In [ ]:
%%sql

SELECT *
FROM mildsalmon_su.channel_month_gross_revenue
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Naver', 237, 11, 0.0464135021097046, 867, 844),
 ('2019-05', 'Google', 253, 10, 0.0395256916996047, 580, 580),
 ('2019-05', 'Instagram', 234, 11, 0.047008547008547, 959, 770),
 ('2019-06', 'Organic', 416, 14, 0.0336538461538462, 1129, 940),
 ('2019-06', 'Instagram', 410, 21, 0.0512195121951219, 1462, 1418),
 ('2019-05', 'Facebook', 247, 14, 0.0566801619433198, 1199, 997),
 ('2019-05', 'Organic', 238, 18, 0.0756302521008403, 1846, 1571),
 ('2019-05', 'Youtube', 244, 10, 0.040983606557377, 529, 529),
 ('2019-06', 'Google', 412, 13, 0.0315533980582524, 947, 947),
 ('2019-06', 'Naver', 398, 16, 0.0402010050251256, 1090, 1090)]

### b. 강사님 풀이

In [4]:
%%sql

SELECT LEFT(ts, 7) "month"
     , usc.channel
     , COUNT(DISTINCT userid) AS uniqueUsers
     , COUNT(DISTINCT (CASE WHEN amount >= 0 THEN userid END)) AS paidUsers
     , ROUND(paidUsers::decimal*100/NULLIF(uniqueUsers, 0), 2) AS conversionRate
     , SUM(amount) AS grossRevenue
     , SUM(CASE WHEN refunded is not True THEN amount END) AS netRevenue
FROM raw_data.user_session_channel usc
LEFT JOIN raw_data.session_timestamp t ON t.sessionid = usc.sessionid
LEFT JOIN raw_data.session_transaction st ON st.sessionid = usc.sessionid
GROUP BY 1, 2
ORDER BY 1, 2
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
42 rows affected.


[('2019-05', 'Facebook', 247, 14, Decimal('5.67'), 1199, 997),
 ('2019-05', 'Google', 253, 10, Decimal('3.95'), 580, 580),
 ('2019-05', 'Instagram', 234, 11, Decimal('4.70'), 959, 770),
 ('2019-05', 'Naver', 237, 11, Decimal('4.64'), 867, 844),
 ('2019-05', 'Organic', 238, 17, Decimal('7.14'), 1846, 1571),
 ('2019-05', 'Youtube', 244, 10, Decimal('4.10'), 529, 529),
 ('2019-06', 'Facebook', 414, 22, Decimal('5.31'), 1578, 1578),
 ('2019-06', 'Google', 412, 13, Decimal('3.16'), 947, 947),
 ('2019-06', 'Instagram', 410, 21, Decimal('5.12'), 1462, 1418),
 ('2019-06', 'Naver', 398, 15, Decimal('3.77'), 1090, 1090),
 ('2019-06', 'Organic', 416, 14, Decimal('3.37'), 1129, 940),
 ('2019-06', 'Youtube', 400, 17, Decimal('4.25'), 1042, 1042),
 ('2019-07', 'Facebook', 558, 32, Decimal('5.73'), 2222, 2144),
 ('2019-07', 'Google', 556, 21, Decimal('3.78'), 1558, 1385),
 ('2019-07', 'Instagram', 567, 25, Decimal('4.41'), 1896, 1766),
 ('2019-07', 'Naver', 553, 19, Decimal('3.44'), 1547, 1547),
 ('2019-07', 'Organic', 557, 24, Decimal('4.31'), 1600, 1600),
 ('2019-07', 'Youtube', 564, 36, Decimal('6.38'), 2210, 2037),
 ('2019-08', 'Facebook', 611, 18, Decimal('2.95'), 1009, 1009),
 ('2019-08', 'Google', 610, 28, Decimal('4.59'), 2210, 1894),
 ('2019-08', 'Instagram', 621, 28, Decimal('4.51'), 2129, 2001),
 ('2019-08', 'Naver', 626, 22, Decimal('3.51'), 1829, 1551),
 ('2019-08', 'Organic', 608, 26, Decimal('4.28'), 1643, 1606),
 ('2019-08', 'Youtube', 614, 18, Decimal('2.93'), 987, 950),
 ('2019-09', 'Facebook', 597, 27, Decimal('4.52'), 2270, 2270),
 ('2019-09', 'Google', 599, 25, Decimal('4.17'), 1872, 1691),
 ('2019-09', 'Instagram', 588, 20, Decimal('3.40'), 1260, 1122),
 ('2019-09', 'Naver', 592, 21, Decimal('3.55'), 1996, 1996),
 ('2019-09', 'Organic', 592, 23, Decimal('3.89'), 1267, 1267),
 ('2019-09', 'Youtube', 588, 15, Decimal('2.55'), 1301, 1301),
 ('2019-10', 'Facebook', 698, 29, Decimal('4.15'), 1650, 1641),
 ('2019-10', 'Google', 699, 30, Decimal('4.29'), 2150, 2098),
 ('2019-10', 'Instagram', 707, 34, Decimal('4.81'), 2568, 2395),
 ('2019-10', 'Naver', 713, 32, Decimal('4.49'), 2695, 2695),
 ('2019-10', 'Organic', 709, 31, Decimal('4.37'), 2762, 2608),
 ('2019-10', 'Youtube', 705, 34, Decimal('4.82'), 2492, 2319),
 ('2019-11', 'Facebook', 688, 26, Decimal('3.78'), 1678, 1678),
 ('2019-11', 'Google', 688, 26, Decimal('3.78'), 2286, 2235),
 ('2019-11', 'Instagram', 669, 25, Decimal('3.74'), 2116, 2116),
 ('2019-11', 'Naver', 667, 26, Decimal('3.90'), 2234, 1987),
 ('2019-11', 'Organic', 677, 34, Decimal('5.02'), 2626, 2255),
 ('2019-11', 'Youtube', 677, 46, Decimal('6.79'), 3532, 3331)]

## B. 검증

### a. Net revenue

Refund가 빠진 Net revenue를 구하기 위해서는 gross revenue - refund을 구해야 하는데, 한번에 구할 방법을 모르겠어서 따로 구함.

gross revenue - refund를 하면 위의 SQL문에서 구한 값과 동일하다.

월별 채널별 refund의 합

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , SUM(C.amount) AS refund
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
GROUP BY 1, 2
ORDER BY 1 ASC
LIMIT 10
;


 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', 202),
 ('2019-05', 'Naver', 23),
 ('2019-05', 'Organic', 275),
 ('2019-05', 'Instagram', 189),
 ('2019-06', 'Organic', 189),
 ('2019-06', 'Instagram', 44),
 ('2019-07', 'Youtube', 173),
 ('2019-07', 'Instagram', 130),
 ('2019-07', 'Facebook', 78),
 ('2019-07', 'Google', 173)]

월별 채널별 판매 총액

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , SUM(C.amount) AS grossRevenue
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
GROUP BY 1, 2
ORDER BY 1 ASC
LIMIT 10
;


 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', 1199),
 ('2019-05', 'Naver', 867),
 ('2019-05', 'Organic', 1846),
 ('2019-05', 'Google', 580),
 ('2019-05', 'Youtube', 529),
 ('2019-05', 'Instagram', 959),
 ('2019-06', 'Google', 947),
 ('2019-06', 'Organic', 1129),
 ('2019-06', 'Naver', 1090),
 ('2019-06', 'Instagram', 1462)]

### b. refunded의 종류 select

처음 Net revenue를 구하기 위해 접근한 방식이 아래와 같다.

```sql
CASE
  WHEN COUNT(D.refund) > 0 THEN ~~
  ELSE
END AS netRevenue
```

그런데 풀다보니 refund가 false인 경우와 true인 경우를 구분할 수 없어서 refunded를 확인해봤다.

refunded는 `TRUE`, `FALSE`, `NULL`이라서 `COUNT()`를 쓰면 TRUE와 FALSE를 구분할 수 없는게 맞다.

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(C.refunded) OVER(PARTITION BY LEFT(A.ts, 7), B.channel, C.refunded)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', True, 3),
 ('2019-05', 'Instagram', True, 2),
 ('2019-05', 'Organic', True, 2),
 ('2019-05', 'Naver', True, 1),
 ('2019-06', 'Instagram', True, 1),
 ('2019-06', 'Organic', True, 2),
 ('2019-07', 'Instagram', True, 1),
 ('2019-07', 'Google', True, 2),
 ('2019-07', 'Youtube', True, 2),
 ('2019-07', 'Facebook', True, 1)]

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(C.refunded) OVER(PARTITION BY LEFT(A.ts, 7), B.channel, C.refunded)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS FALSE
ORDER BY 1 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', False, 11),
 ('2019-05', 'Naver', False, 10),
 ('2019-05', 'Organic', False, 16),
 ('2019-05', 'Youtube', False, 10),
 ('2019-05', 'Instagram', False, 9),
 ('2019-05', 'Google', False, 10),
 ('2019-06', 'Naver', False, 16),
 ('2019-06', 'Google', False, 13),
 ('2019-06', 'Organic', False, 12),
 ('2019-06', 'Instagram', False, 20)]

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS channel
      , C.refunded
      , COUNT(*) OVER(PARTITION BY LEFT(A.ts, 7), B.channel, C.refunded)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS NULL
ORDER BY 1 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', None, 1089),
 ('2019-05', 'Naver', None, 1054),
 ('2019-05', 'Google', None, 1031),
 ('2019-05', 'Organic', None, 971),
 ('2019-05', 'Instagram', None, 993),
 ('2019-05', 'Youtube', None, 1049),
 ('2019-06', 'Naver', None, 1716),
 ('2019-06', 'Youtube', None, 1754),
 ('2019-06', 'Facebook', None, 1685),
 ('2019-06', 'Organic', None, 1744)]

### c. paidUsers 검증

|yearmonth|	channel|	paidusers|	
|---|---|---|
|2019-05|	Facebook|	14|	
|2019-05|	Naver	|	11|	
|2019-05|	Organic|	18|


In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
     , B.channel AS user_session_channel
     , C.amount
     , C.refunded
     , COUNT(C.amount) OVER(PARTITION BY LEFT(A.ts, 7), B.channel)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.amount IS NOT NULL
ORDER BY 1, 2, 3 ASC
LIMIT 100
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
100 rows affected.


[('2019-05', 'Facebook', 3, True, 14),
 ('2019-05', 'Facebook', 16, False, 14),
 ('2019-05', 'Facebook', 29, False, 14),
 ('2019-05', 'Facebook', 43, False, 14),
 ('2019-05', 'Facebook', 63, False, 14),
 ('2019-05', 'Facebook', 80, True, 14),
 ('2019-05', 'Facebook', 105, False, 14),
 ('2019-05', 'Facebook', 107, False, 14),
 ('2019-05', 'Facebook', 111, False, 14),
 ('2019-05', 'Facebook', 119, True, 14),
 ('2019-05', 'Facebook', 127, False, 14),
 ('2019-05', 'Facebook', 129, False, 14),
 ('2019-05', 'Facebook', 129, False, 14),
 ('2019-05', 'Facebook', 138, False, 14),
 ('2019-05', 'Google', 2, False, 10),
 ('2019-05', 'Google', 9, False, 10),
 ('2019-05', 'Google', 16, False, 10),
 ('2019-05', 'Google', 37, False, 10),
 ('2019-05', 'Google', 48, False, 10),
 ('2019-05', 'Google', 69, False, 10),
 ('2019-05', 'Google', 91, False, 10),
 ('2019-05', 'Google', 92, False, 10),
 ('2019-05', 'Google', 102, False, 10),
 ('2019-05', 'Google', 114, False, 10),
 ('2019-05', 'Instagram', 15, False, 11),
 ('2019-05', 'Instagram', 74, False, 11),
 ('2019-05', 'Instagram', 75, True, 11),
 ('2019-05', 'Instagram', 84, False, 11),
 ('2019-05', 'Instagram', 85, False, 11),
 ('2019-05', 'Instagram', 97, False, 11),
 ('2019-05', 'Instagram', 99, False, 11),
 ('2019-05', 'Instagram', 103, False, 11),
 ('2019-05', 'Instagram', 104, False, 11),
 ('2019-05', 'Instagram', 109, False, 11),
 ('2019-05', 'Instagram', 114, True, 11),
 ('2019-05', 'Naver', 7, False, 11),
 ('2019-05', 'Naver', 20, False, 11),
 ('2019-05', 'Naver', 23, True, 11),
 ('2019-05', 'Naver', 47, False, 11),
 ('2019-05', 'Naver', 60, False, 11),
 ('2019-05', 'Naver', 88, False, 11),
 ('2019-05', 'Naver', 98, False, 11),
 ('2019-05', 'Naver', 124, False, 11),
 ('2019-05', 'Naver', 126, False, 11),
 ('2019-05', 'Naver', 131, False, 11),
 ('2019-05', 'Naver', 143, False, 11),
 ('2019-05', 'Organic', 23, False, 18),
 ('2019-05', 'Organic', 45, False, 18),
 ('2019-05', 'Organic', 57, False, 18),
 ('2019-05', 'Organic', 58, False, 18),
 ('2019-05', 'Organic', 60, False, 18),
 ('2019-05', 'Organic', 84, False, 18),
 ('2019-05', 'Organic', 92, False, 18),
 ('2019-05', 'Organic', 110, False, 18),
 ('2019-05', 'Organic', 114, False, 18),
 ('2019-05', 'Organic', 114, False, 18),
 ('2019-05', 'Organic', 121, False, 18),
 ('2019-05', 'Organic', 127, True, 18),
 ('2019-05', 'Organic', 131, False, 18),
 ('2019-05', 'Organic', 135, False, 18),
 ('2019-05', 'Organic', 139, False, 18),
 ('2019-05', 'Organic', 140, False, 18),
 ('2019-05', 'Organic', 148, True, 18),
 ('2019-05', 'Organic', 148, False, 18),
 ('2019-05', 'Youtube', 0, False, 10),
 ('2019-05', 'Youtube', 2, False, 10),
 ('2019-05', 'Youtube', 26, False, 10),
 ('2019-05', 'Youtube', 34, False, 10),
 ('2019-05', 'Youtube', 43, False, 10),
 ('2019-05', 'Youtube', 49, False, 10),
 ('2019-05', 'Youtube', 74, False, 10),
 ('2019-05', 'Youtube', 87, False, 10),
 ('2019-05', 'Youtube', 106, False, 10),
 ('2019-05', 'Youtube', 108, False, 10),
 ('2019-06', 'Facebook', 6, False, 22),
 ('2019-06', 'Facebook', 7, False, 22),
 ('2019-06', 'Facebook', 9, False, 22),
 ('2019-06', 'Facebook', 19, False, 22),
 ('2019-06', 'Facebook', 19, False, 22),
 ('2019-06', 'Facebook', 45, False, 22),
 ('2019-06', 'Facebook', 46, False, 22),
 ('2019-06', 'Facebook', 50, False, 22),
 ('2019-06', 'Facebook', 51, False, 22),
 ('2019-06', 'Facebook', 64, False, 22),
 ('2019-06', 'Facebook', 72, False, 22),
 ('2019-06', 'Facebook', 81, False, 22),
 ('2019-06', 'Facebook', 85, False, 22),
 ('2019-06', 'Facebook', 87, False, 22),
 ('2019-06', 'Facebook', 89, False, 22),
 ('2019-06', 'Facebook', 101, False, 22),
 ('2019-06', 'Facebook', 118, False, 22),
 ('2019-06', 'Facebook', 122, False, 22),
 ('2019-06', 'Facebook', 122, False, 22),
 ('2019-06', 'Facebook', 125, False, 22),
 ('2019-06', 'Facebook', 127, False, 22),
 ('2019-06', 'Facebook', 133, False, 22),
 ('2019-06', 'Google', 3, False, 13),
 ('2019-06', 'Google', 9, False, 13),
 ('2019-06', 'Google', 9, False, 13),

### d. uniqueUsers 검증

|yearmonth|	channel|	uniqueusers|		
|---|---|---|
|2019-05|	Facebook	|247|
|2019-05|	Naver|	237|
|2019-05|	Organic|	238|
|2019-05|	Google|	253|
|2019-05|	Youtube|	244|

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
     , B.channel AS user_session_channel
     , COUNT(B.userid) AS Users
     , COUNT(DISTINCT B.userid) AS uniqueUsers
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE LEFT(A.ts, 7) = '2019-05'
  AND B.channel IN ('Youtube', 'Facebook', 'Naver', 'Organic', 'Google')
GROUP BY 1, 2
ORDER BY 1 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
5 rows affected.


[('2019-05', 'Facebook', 1103, 247),
 ('2019-05', 'Naver', 1065, 237),
 ('2019-05', 'Organic', 989, 238),
 ('2019-05', 'Youtube', 1059, 244),
 ('2019-05', 'Google', 1041, 253)]

# ETC. 새로 알게된 것


## A. Partition by에서의 aggregation 표기

WINDOW 함수에서 PARTITION BY 구문에 1, 2, 3과 같은 표기는 정상적으로 동작하지 않는다.

오라클과 다르게 GROUP BY절에서 1, 2, 3을 통해 컬럼을 간편하게 지정할 수 있길래 PARTITION BY 구문에서도 사용해봤는데, 데이터를 자세히 들여다보니까 이상한 값이 나와있다.

---

단순하게 비교해봐도 아래 두 쿼리는 `PARTITION BY`절만 다르고 전부 동일하다. 그리고 `PARTITION BY`절도 동일한 의미로 작성했으므로 동일한 결과가 도출되어야 한다.

그러나 출력된 결과는 다르다.

`PARTITION BY 1, 2, 3`은 단순히 모든 row를 나타내는 듯 하다.

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS ch
      , C.refunded AS refund
      , COUNT(C.refunded) OVER(PARTITION BY LEFT(A.ts, 7), B.channel, C.refunded)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1, 2 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', True, 3),
 ('2019-05', 'Instagram', True, 2),
 ('2019-05', 'Naver', True, 1),
 ('2019-05', 'Organic', True, 2),
 ('2019-06', 'Instagram', True, 1),
 ('2019-06', 'Organic', True, 2),
 ('2019-07', 'Facebook', True, 1),
 ('2019-07', 'Google', True, 2),
 ('2019-07', 'Instagram', True, 1),
 ('2019-07', 'Youtube', True, 2)]

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS ch
      , C.refunded AS refund
      , COUNT(C.refunded) OVER(PARTITION BY 1, 2, 3)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1, 2 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', True, 54),
 ('2019-05', 'Instagram', True, 54),
 ('2019-05', 'Naver', True, 54),
 ('2019-05', 'Organic', True, 54),
 ('2019-06', 'Instagram', True, 54),
 ('2019-06', 'Organic', True, 54),
 ('2019-07', 'Facebook', True, 54),
 ('2019-07', 'Google', True, 54),
 ('2019-07', 'Instagram', True, 54),
 ('2019-07', 'Youtube', True, 54)]

아래에 `DISTINCT`하지 않은 결과를 보면 Facebook 3개, Instagram 2개, Naver 1개 등을 보여준다.

In [ ]:
%%sql

SELECT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS ch
      , C.refunded AS refund
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1, 2 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', True),
 ('2019-05', 'Facebook', True),
 ('2019-05', 'Facebook', True),
 ('2019-05', 'Instagram', True),
 ('2019-05', 'Instagram', True),
 ('2019-05', 'Naver', True),
 ('2019-05', 'Organic', True),
 ('2019-05', 'Organic', True),
 ('2019-06', 'Instagram', True),
 ('2019-06', 'Organic', True)]

### a. 그럼 별칭은 PARTITION BY에서 적용될까?

별칭은 적용된다.

신기한 SQL의 세계

In [ ]:
%%sql

SELECT DISTINCT LEFT(A.ts, 7) AS yearMonth
      , B.channel AS ch
      , C.refunded AS refund
      , COUNT(C.refunded) OVER(PARTITION BY yearMonth, ch, refund)
FROM raw_data.session_timestamp A
JOIN raw_data.user_session_channel B ON A.sessionid = B.sessionid
LEFT JOIN raw_data.session_transaction C ON B.sessionid = C.sessionid
WHERE C.refunded IS TRUE
ORDER BY 1, 2 ASC
LIMIT 10
;

 * postgresql://mildsalmon_su:***@learnde.cduaw970ssvt.ap-northeast-2.redshift.amazonaws.com:5439/dev
10 rows affected.


[('2019-05', 'Facebook', True, 3),
 ('2019-05', 'Instagram', True, 2),
 ('2019-05', 'Naver', True, 1),
 ('2019-05', 'Organic', True, 2),
 ('2019-06', 'Instagram', True, 1),
 ('2019-06', 'Organic', True, 2),
 ('2019-07', 'Facebook', True, 1),
 ('2019-07', 'Google', True, 2),
 ('2019-07', 'Instagram', True, 1),
 ('2019-07', 'Youtube', True, 2)]